core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala

```diff
@@ -0,0 +1,202 @@
+
+    val updateMetadataRequest = new UpdateMetadataRequest(controllerId, controllerEpoch, partitionStates.asJava, brokers.asJava)
+    cache.updateCache(15, updateMetadataRequest)
+
+    val topicMetadatas = cache.getTopicMetadata(Set(topic), SecurityProtocol.PLAINTEXT)
+    assertEquals(1, topicMetadatas.size)
+
+    val topicMetadata = topicMetadatas.head
+    assertEquals(Errors.NONE, topicMetadata.error())
+
+    val partitionMetadatas = topicMetadata.partitionMetadata()
+    assertEquals(1, partitionMetadatas.size)
+
+    val partitionMetadata = partitionMetadatas.get(0)
+    assertEquals(0, partitionMetadata.partition())
+    assertEquals(Errors.LEADER_NOT_AVAILABLE, partitionMetadata.error())
+    assertTrue(partitionMetadata.isr().isEmpty)
+    assertEquals(1, partitionMetadata.replicas().size())
+    assertEquals(0, partitionMetadata.replicas().get(0).id())
+  }
+
+  @Test
+  def getTopicMetadataReplicaNotAvailable() {
+    val topic = "topic"
+
+    val cache = new MetadataCache(1)
+
+    val zkVersion = 3
+    val controllerId = 2
+    val controllerEpoch = 1
+    val brokers = Set(new Broker(0, Map(SecurityProtocol.PLAINTEXT -> new EndPoint("foo", 9092)).asJava))
+
+    // replica 1 is not available
+    val leader = 0
+    val leaderEpoch = 0
+    val replicas: util.Set[java.lang.Integer] = asSet(0, 1)
+    val isr: util.List[java.lang.Integer] = asList(0)
+
+    val partitionStates = Map(
+      new TopicPartition(topic, 0) -> new PartitionState(controllerEpoch, leader, leaderEpoch, isr, zkVersion, replicas))
+
+    val updateMetadataRequest = new UpdateMetadataRequest(controllerId, controllerEpoch, partitionStates.asJava, brokers.asJava)
+    cache.updateCache(15, updateMetadataRequest)
+
+    val topicMetadatas = cache.getTopicMetadata(Set(topic), SecurityProtocol.PLAINTEXT)
+    assertEquals(1, topicMetadatas.size)
+
+    val topicMetadata = topicMetadatas.head
+    assertEquals(Errors.NONE, topicMetadata.error())
+
+    val partitionMetadatas = topicMetadata.partitionMetadata()
+    assertEquals(1, partitionMetadatas.size)
+
+    val partitionMetadata = partitionMetadatas.get(0)
+    assertEquals(0, partitionMetadata.partition())
+    assertEquals(Errors.REPLICA_NOT_AVAILABLE, partitionMetadata.error())
+    assertEquals(Set(0), partitionMetadata.replicas().asScala.map(_.id()).toSet)
+    assertEquals(Set(0), partitionMetadata.isr().asScala.map(_.id()).toSet)
+  }
+
+  @Test
+  def getTopicMetadataIsrNotAvailable() {
+    val topic = "topic"
+
+    val cache = new MetadataCache(1)
+
+    val zkVersion = 3
+    val controllerId = 2
+    val controllerEpoch = 1
+    val brokers = Set(new Broker(0, Map(SecurityProtocol.PLAINTEXT -> new EndPoint("foo", 9092)).asJava))
+
+    // replica 1 is not available
+    val leader = 0
+    val leaderEpoch = 0
+    val replicas: util.Set[java.lang.Integer] = asSet(0)
+    val isr: util.List[java.lang.Integer] = asList(0, 1)
```

Comment:
Nitpick: I think it's a bit nicer if we avoid the type on the left-hand side and use eg `asSet[Integer](0)`.